tests: kernel: mutex: move to new ztest API

smrtos opened the pull request:
Move below tests to new ztest API:

- tests/kernel/mutex/mutex_api
- tests/kernel/mutex/mutex_error_case
- tests/kernel/mutex/syst_mutex


Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos committed:
tests: kernel: mutex_api: move to new ztest API

Move mutex_api tests to new ztest API.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos committed:
tests: kernel: mutex_error_case: move to new ztest API

Move mutex_error_case tests to new ztest API.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos commented:
New ztest fx doesn't have a logic to launch static threads. For test suites that use one-shot static threads, when they are shuffled and executed repeatedly, some later executions won't have a living static thread to interact with. And tests can fail. Such as the tests/kernel/mutex/sys_mutex in this PR.  Current way is to use dynamic thread instead.

New ztest assert if 1cpu operation takes too long. This can cause some false failures. Maybe we can use a warning instead of assert?
https://github.com/zephyrproject-rtos/zephyr/blob/d77389464b92345f62bc8c40c752cbd48165ecb3/subsys/testsuite/ztest/src/ztest_new.c#L137

smrtos commented:
Have to add the `CONFIG_MP_NUM_CPUS=1` back to the `mutex_api` test, though it is merely to make below new ztest assert happy.

https://github.com/zephyrproject-rtos/zephyr/blob/d77389464b92345f62bc8c40c752cbd48165ecb3/subsys/testsuite/ztest/src/ztest_new.c#L137

chen-png commented on tests/kernel/mutex/sys_mutex/src/thread_competition.c:
```diff
@@ -48,9 +45,8 @@ static void high_prio_t1_wait_for_mutex(void *p1, void *p2, void *p3)
 	sys_mutex_lock(pmutex, K_FOREVER);
 
-	flag[index] = HIGH_T1;
+	flag[0] = HIGH_T1;
```
with this change, when you invoke assert to check flag[0] later, this only guarantees that this thread handler function got executed, it can't guarantee it was the first one to be executed.
the goal of this test, when we unlock a mutex, it should be given to the highest priority and longest waiting time in the mutex's waiting queue.

chen-png commented on tests/kernel/mutex/sys_mutex/src/main.c:
```diff
@@ -67,7 +67,33 @@ static SYS_MUTEX_DEFINE(no_access_mutex);
+#define DEFINE_PARTICIPANT_THREAD(id)                               \
+		K_THREAD_STACK_DEFINE(thread_##id##_stack_area, STACKSIZE); \
+		struct k_thread thread_##id##_thread_data;                  \
+		k_tid_t thread_##id##_tid;
```
this var is unused

andyross reviewed (changes requested):
Could you take the text you gave as an explanation in the github comment and add it to the commit message?  Right now the commit just says "use new API", but the code change looks like "**completely rewrite test**".

Even better would be to do it in two stages, where the first commit is "use dynamic threads" and the second is "use new ztest API".

smrtos committed:
tests: Use dynamic thread instead of static thread for sys_mutex test

As of now, both the old and new ztest framework don't have a logic to
launch static threads. Static threads can only be launched by Zephyr
itself during boot.  If a test interact with static threads, when it
is executed multiple times without reboot, only the first few runs can
have static threads to interact with. After the static threads exit,
later runs will have nothing to interact with, which can lead to failure.
One unfortunate example is the sys_mutex test.
This commit replaces static threads with dynamic thread to make the test
repeatable. In future, maybe static threads launching logic should  be
covered by the ztest framework so richer scenarios can be tested.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos committed:
tests: Enhance the thread_competition case in sys_mutex test

Current thread_competition() case cannot verify the order of mutex
taking. This commit enhances on that.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos committed:
tests: kernel: syst_mutex: move to new ztest API

Move sys_mutex tests to new ztest API.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos commented:
> Could you take the text you gave as an explanation in the github comment and add it to the commit message? Right now the commit just says "use new API", but the code change looks like "**completely rewrite test**".
> 
> Even better would be to do it in two stages, where the first commit is "use dynamic threads" and the second is "use new ztest API".

Yes, indeed.

I have split the PR into more commits. And added more detailed commit msg. Thanks for the advice.

smrtos commented:
@andyross  Could you help review this one?

enjiamai commented on tests/kernel/mutex/mutex_api/src/test_mutex_apis.c:
```diff
@@ -386,21 +387,16 @@ void test_mutex_priority_inheritance(void)
-void test_main(void)
+static void *mutex_api_tests_setup(void)
 {
+#ifdef CONFIG_USERSPACE
```
I think we can remove the #ifdef  because when CONFIG_USERSPACE=n, k_thread_access_grant will do nothing.

smrtos replied:
Maybe this can make things a bit explicit.

andyross reviewed (approved):
Looks much clearer, thanks
